src/lib.rs

```diff
@@ -74,49 +90,242 @@
+//! RUSTFLAGS="--cfg loom" cargo test --test loom_my_struct --release
+//! ```
 //!
-//! For more usage details, see the [README]
+//! Note that you will generally want to run loom tests with `--release` since loom must execute
+//! each test closure a large number of times, at which point the speed win from optimized code
+//! makes a big difference.
+//!
+//! # Debugging Loom Failures
+//!
+//! Loom's deterministic execution allows the specific chain of events leading to a test failure
+//! can be isolated for debugging. When a loom test fails, the first step is to isolate the exact
+//! execution path that resulted in the failure. To do this, Loom is able to output the execution
+//! path to a file. Two environment variables are useful for this process:
+//!
+//! - `LOOM_CHECKPOINT_FILE`
+//! - `LOOM_CHECKPOINT_INTERVAL`
+//!
+//! The first specifies the file to write to and read from. The second specifies how often to write
+//! to the file. If the execution fails on the 10,000,000th permutation, it is faster to write to a
+//! file every 10,000 iterations instead of every single one.
+//!
+//! To isolate the exact failing path, first run the following command to generate the checkpoint
+//! for the failing scenario:
+//!
+//! ```console
+//! LOOM_CHECKPOINT_FILE=my_test.json [other env vars] \
+//!     cargo test --test loom_my_struct --release [failing test]
+//! ```
 //!
-//! [README]: https://github.com/tokio-rs/loom
+//! Then this to check that the next permutation indeed triggers the fault:
+//!
+//! ```console
+//! LOOM_CHECKPOINT_INTERVAL=1 LOOM_CHECKPOINT_FILE=my_test.json [other env vars] \
+//!     cargo test --test loom_my_struct --release [failing test]
+//! ```
+//!
+//! The test should fail on the first permutation, effectively isolating the failure
+//! scenario.
+//!
+//! The next step is to enable additional log output for just the failing permutation. Again, there
+//! are some environment variables for this:
+//!
+//! - `LOOM_LOG`
+//! - `LOOM_LOCATION`
+//!
+//! The first environment variable, `LOOM_LOG`, outputs a marker on every thread switch. This helps
+//! with tracing the exact steps in a threaded environment that results in the test failure.
+//!
+//! The second, `LOOM_LOCATION`, enables location tracking. This includes additional information in
+//! panic messages that helps identify which specific field resulted in the error.
+//!
+//! Put together, the command becomes (yes, we know this is not great... but it works):
+//!
+//! ```console
+//! LOOM_LOG=1 \
+//!     LOOM_LOCATION=1 \
+//!     LOOM_CHECKPOINT_INTERVAL=1 \
+//!     LOOM_CHECKPOINT_FILE=my_test.json \
+//!     RUSTFLAGS="--cfg loom" \
+//!     [other env vars] \
+//!     cargo test --test loom_my_struct --release [failing test]
+//! ```
+//!
+//! This should provide you with a trace of all the concurrency events leading up to the failure,
+//! which should allow you to identify how the bug is triggered.
+//!
+//! # Limitations and Caveats
```

Comment:
It's also probably worth adding a section on loom being an "intrusive" tool requiring the user to use loom's APIs. The implication being that loom cannot catch bugs that are from code *not* instrumented. `tsan` does not have this problem as it is not intrusive.